[system] Fix breakpoints typescript types

Kujawadl opened the pull request:
# Changes
Box prop types previously didn't allow for responsive values (e.g. arrays/objects) and also did not account for variables in `fontWeight`.

`fontWeight` is now a `string | number` value to accept variable names without compiler errors, as opposed to the union type defined in React.CSSProperties.

# Necessity
Without these changes, the following uses in typescript would have compile errors:

```tsx
{/* "fontWeightBold" is not a valid value of React.CSSProperties["fontWeight"] */}
<Box fontWeight="fontWeightBold">
  ...
</Box>

{/* Type { ... } is not assignable to type string | number */}
<Box m={{ xs: 1, sm: 2, md: 3 }}>
...
</Box>
```

There are probably other props that need updating, but fontWeight was the only one that I found that seems to have variables that would actually be used. The others are:

```
flexDirection
flexGrow
flexShrink
flexWrap
overflowX
overflowY
position
textAlign
zIndex
```

All of these have union types, e.g. `"column" | "column-reverse" | "row" | "row-reverse" | ...`, and I can't think of any variables that would be used, so it's probably better to keep the union types than lose them for the sake of variables that aren't likely to be used.

# Alternative Options

I would have just created a new union type for fontWeight, something like:

```tsx
fontWeight?:
  | React.CSSProperties['fontWeight']
  | 'fontWeightLight'
  | 'fontWeightRegular' 
  | ...
```

Except that would prevent users from using custom variables, e.g. in our codebase we have a `fontWeightExtraBold`. The solution to that would be to create and export a named type, e.g. `FontWeightValue` and set it to the above. I think that's the most complete solution, but starts to get tedious to maintain, so I opted for simple here.

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->
# <!-- Blank H1 for the horizontal rule -->
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/next/CONTRIBUTING.md#submitting-a-pull-request).

eps1lon commented on packages/material-ui/src/Box/Box.d.ts:
```diff
@@ -1,9 +1,27 @@
+  | [Value, Value]
+  | [Value, Value, Value]
+  | [Value, Value, Value, Value]
+  | [Value, Value, Value, Value, Value]
```
Is the union of tuples required or can we just use `Value[]`?

Kujawadl replied:
We could just use `Value[]` but I preferred using tuples as it limits the number of values you can provide from 2 to 5, which is more accurate.

oliviertassinari replied:
People should be able to use custom breakpoints, e.g. small, medium, large, or mobile, table, desktop. I prefer @eps1lon proposal.

Kujawadl replied:
So we need to be able to support more than 5 breakpoints?

Kujawadl commented:
Okay, a few more changes here, that was quite the rabbit hole to go down. I started converting docs demos to typescript (for testing) and found that there were some more type changes that needed to be made and that weren't being tested.

Some notes about the recent changes:

* My eslint plugin was throwing errors about jsx in typescript, so I added .tsx files to the acceptable files list. This doesn't affect anything practically, but silences IDE warnings for anybody who's using an IDE with eslint integration. I can revert this if you'd prefer, I just figured it's useful for anybody else in my situation.
* The `breakpoints` function in `material-ui-system` was typed to accept a style function whose props already has the breakpoints, and then return the same type of style function. It should be that it takes a non-breakpoint-related style function (e.g. `typography`) and returns a style function with the breakpoints added in. I had to fix this to get the collocation doc demo working without tslint errors.
* I added a link in the `ResponsiveValue` type's jsdoc annotation to the responsive page that shows demos of these responsive props in action. If you don't want jsdoc annotations, I can revert this commit.

oliviertassinari commented on docs/src/pages/system/basics/basics.md:
````diff
@@ -139,94 +139,21 @@ If you are already using `@material-ui/core`, we encourage you to start with the
-  },
-};
-```
+{{"demo": "pages/system/basics/ResponsiveValues.js", "defaultCodeOpen": true}}
````
Do we need a demo here? Shouldn't we move the TypeScript code as a test case instead?

eps1lon replied:
Do we have this documented somewhere else?

oliviertassinari replied:
To be clear, we need to document the API as well as to make sure that it works with TypeScript. I'm not sure it's two concerns we should merge. My suggestion is that we keep the markdown but we move the live version to a TypeScript test file.

Kujawadl replied:
There doesn't appear to by anything in place for a typescript test suite; it was my understanding that these typescript demos are the tests. Should I just restore the js markdown (so we don't have the "demo" in the docs) but leave the `ResponsiveValues.tsx` file so it can be "tested"?

If I do that, and intentionally screw up something with the ResponsiveValues "demo", the `yarn typescript` command fails like it should, and we don't have to have the export because it's not being used as a demo.

Once I get an answer on this, I'll commit & push everything that was requested in this round of review.

oliviertassinari replied:
We do have dedicated TypeScript tests. Demos are not the tests but they help with the coverage. I have no problem with a live demo, as long as it's optimized for learning, not testing.

Kujawadl replied:
Oh well in that case I'll probably leave it as-is. I originally just copy/pasted from the markdown into a typescript file and added types for reference. 

The only reason it has a default export is because the demo component appears to require an export. It doesn't render anything, it just shows the code.

Other than that, the docs for this section are unchanged, so we get the original code with a typescript variant + coverage from the tsx file.

oliviertassinari replied:
If it just show the code, it should be in the markdown, not in the runtime.

Kujawadl replied:
Okay, I'll move it back to the markdown. Can you help me find the typescript tests? I'm struggling to find them at the moment.

Kujawadl replied:
I moved it back into the markdown and removed the demo file for this. I don't think it needs extra testing, as the theme stuff already has extensive tests in `packages/material-ui-styles/test/*.spec.tsx`.

oliviertassinari commented on docs/src/pages/system/basics/basics.md:
```diff
@@ -139,94 +139,21 @@ If you are already using `@material-ui/core`, we encourage you to start with the
-<Box p={[2, 3, 4]} />
-
-/**
- * Outputs:
```
I'm wondering if we shouldn't add the CSS output information back.

oliviertassinari commented on docs/src/pages/system/basics/ArrayApi.js:
```diff
@@ -0,0 +1,8 @@
+import Box from '@material-ui/core/Box';
+
+function ArrayApi() {
+  return <Box fontSize={[12, 18, 24, 36, 72]}>Array API</Box>;
```
I think that 3 breakpoints are enough to make our point, same for the other demos:
```suggestion
  return <Box fontSize={[12, 18, 24]}>Array API</Box>;
```

If you want to test 5 breakpoints, a test case would be better than leaking it to the educational content that the demos.

eps1lon commented on docs/src/pages/system/basics/ArrayApi.js:
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
```
I don't understand what is different with your proposal.

oliviertassinari replied:
The difference is:
```diff
--- a/docs/src/pages/system/basics/ArrayApi.js
+++ b/docs/src/pages/system/basics/ArrayApi.js
@@ -1,8 +1,6 @@
 import React from 'react';
 import Box from '@material-ui/core/Box';

-function ArrayApi() {
+export default function ArrayApi() {
   return <Box fontSize={[12, 18, 24, 36, 72]}>Array API</Box>;
 }
-
-export default ArrayApi;
```

Kujawadl commented:
Concerning #15548: I don't think that PR addresses the same issues as this one, but if @merceyz doesn't have time to continue working on it I can take a look later today or tomorrow as it's also something my team would find useful. Just let me know.

eps1lon commented:
I'm sorry I forgot this was here. Could you rebase this with master without changing types implementation and see if the demos work? #15884 might've fixed this already.

oliviertassinari commented:
@eps1lon I have rebased the pull request maintaining the failing case:
```jsx
import React from 'react';
import { typography, breakpoints } from '@material-ui/system';
import styled from 'styled-components';

const Box = styled.div`
  ${breakpoints(typography)}
`;

/**
 * Outputs:
 *
 * font-size: 12px;
 * @media (min-width: 600px) {
 *   font-size: 18px;
 * }
 * @media (min-width: 960px) {
 *   font-size: 24px;
 * }
 */
export default function CollocationApi() {
  return (
    <Box xs={{ fontSize: 12 }} sm={{ fontSize: 18 }} md={{ fontSize: 24 }}>
      Collocation API
    </Box>
  );
}
```
